crates/bevy_ecs/src/world/entity_ref.rs

```diff
@@ -375,6 +375,57 @@ impl<'w> EntityMut<'w> {
         self
     }
 
+    /// Inserts a dynamic [`Bundle`] into the entity.
+    ///
+    /// This will overwrite any previous value(s) of the same component type.
+    ///
+    /// You should prefer to use the typed API [`EntityMut::insert`] where possible and only
+    /// use this in cases where there is no static type corresponding to the [`BundleId`].
+    ///
+    /// To obtain a [`BundleId`] you can call [`World::init_bundle`] or [`World::init_dynamic_bundle`].
+    ///
+    /// # Safety
+    /// - Each [`OwningPtr`] must be a valid reference to the type represented by [`ComponentId`]
+    /// - Each storage type must be valid for the given [`ComponentId`]
+    /// - Bundle iterator must have identical element order as is defined by [`BundleId`]
```

Comment:
That's a good point it would be a big ergonomic improvement, will fix.

Reply:
Going to second this. The changes to require passing in `StorageType` on insert was a performance optimization for the non-dynamic use case, since we have it at compile time there. Here, we definitely don't want to expose this safety invariant to normal users: it'd be a fast path to creating trivial UB.